portalnet/src/overlay_service.rs

```diff
@@ -1232,6 +1232,12 @@ where
             OverlayRequestError::AcceptError("unable to find ENR for NodeId".to_string())
         })?;
         let enr = crate::discovery::UtpEnr(node_addr.enr);
+
+        // avoid memory allocation if trace isn't enabled
```

Comment:
I don't love this pattern of adding complexity (if statements) to the core code to handle an edge case of a trace-level log. 
- Our code (especially `overlay_service.rs`) is already quite complex / convoluted. 
- I understand your personal vendetta against unnecessary memory allocation, but imo, this project is not nearly close to the point where we're concerned with optimization. These kinds of improvements are best saved for later down the road when we are optimizing, and can do so in a structured way. ( profiling / identifying / correcting bottlenecks, etc). Imo, at the current stage of this project we should prioritize code simplicity > performance.

Reply:
I am not looking to optimize vaste amounts of memory usage. I am trying to be insightful of when I write new code that is is memory effiencent. If I can avoid tech debt when it comes to memory I think that is important. Little changes add up over time.

I don't want to do premature optimization either, but if we can achieve less memory usage without doing additional work now, it will greatly reduce the problem down the line. I am trying to be proactive about it so the debt doesn't pile up for now reason.

> Our code (especially overlay_service.rs) is already quite complex / convoluted.

If that is the case we need to refactor the code. If we are running offer 100 times avoiding calculating the base64 ENR is worth it in itself without the memory saves.

Reply:
Yeah, there's some gray area here. In this case, I agree that if the only reason we're doing a calculation is to generate the base64 encoding for the log, then it is nice to skip it if the logging is disabled (especially for something as low as a trace that's rarely enabled, and a part of the code that's this regularly called). Skipping log work, especially in hot areas of the code, is a pattern that we ended up having to use pretty often in py-evm to improve performance.

Reply:
Basically, I'm surprised that we would even need to consider this (IMO) antipattern to get the best use out of a very popular rust library. Digging into it slightly - based on [this comment by a maintainer](https://www.reddit.com/r/rust/comments/x9nypb/comment/inusmd5/?utm_source=share&utm_medium=web2x&context=3) it appears to me like if the log level is not enabled, then we don't have to worry about the fields being evaluated. 

> a runtime-disabled span or event compiles down to two instructions: a branch and load of an atomic boolean, which is morally close to nothing on most processors.

But maybe I'm misunderstanding the comment? @KolbyML have you confirmed that we have a runtime performance hit? If it's the case that the runtime hit is "morally close to nothing" then I'd argue that it's not worth the extra lines of code.

Reply:
> Basically, I'm surprised that we would even need to consider this (IMO) antipattern to get the best use out of a very popular rust library. Digging into it slightly - based on [this comment by a maintainer](https://www.reddit.com/r/rust/comments/x9nypb/comment/inusmd5/?utm_source=share&utm_medium=web2x&context=3) it appears to me like if the log level is not enabled, then we don't have to worry about the fields being evaluated.
> 
> > a runtime-disabled span or event compiles down to two instructions: a branch and load of an atomic boolean, which is morally close to nothing on most processors.


I don't understand how what the maintainer said is relevant, I am not complaining about adding ``trace!()`` or ``error!()`` as a problem and my code doesn't signify that. I didn't want to allocate memory for a string and run ``enr.to_base64()`` which encodes to the rlp byte array then encodes to a hex string for a trace which only people debugging will have to run. Even a 0.04% proformance hit will add up over time if you keep making it. Little changes add up a lot over a big codebase.

> But maybe I'm misunderstanding the comment? @KolbyML have you confirmed that we have a runtime performance hit? If it's the case that the runtime hit is "morally close to nothing" then I'd argue that it's not worth the extra lines of code.

I made some changes so it is less intrusive
```rust
        // Generate a connection ID for the uTP connection if there is data we would like to
        // accept.
        let node_addr = self.discovery.cached_node_addr(source).ok_or_else(|| {
            OverlayRequestError::AcceptError("unable to find ENR for NodeId".to_string())
        })?;
        let enr_str = if enabled!(Level::TRACE) {
            node_addr.enr.to_base64()
        } else {
            String::with_capacity(0)
        };
        let enr = crate::discovery::UtpEnr(node_addr.enr);
        let cid = self.utp_socket.cid(enr, false);
        let cid_send = cid.send;
        let validator = Arc::clone(&self.validator);
        let store = Arc::clone(&self.store);
        let kbuckets = Arc::clone(&self.kbuckets);
        let command_tx = self.command_tx.clone();
        let utp = Arc::clone(&self.utp_socket);
        let metrics = self.metrics.clone();
```

^ I was able to make it look nicer which I think is the best solution

``let enr_str = node_addr.enr.to_base64();``

But I will just push this to the branch ^